crypto/pkcs7/pkcs7.c

```diff
@@ -608,3 +609,480 @@ void PKCS7_RECIP_INFO_get0_alg(PKCS7_RECIP_INFO *ri, X509_ALGOR **penc) {
+  OPENSSL_free(ek);
+  return ret;
+}
+
+BIO *PKCS7_dataInit(PKCS7 *p7, BIO *bio) {
+  GUARD_PTR(p7);
+  BIO *out = NULL, *btmp = NULL;
+  const EVP_CIPHER *evp_cipher = NULL;
+  STACK_OF(PKCS7_RECIP_INFO) *rsk = NULL;
+  STACK_OF(X509_ALGOR) *md_sk = NULL;
+  X509_ALGOR *xalg = NULL;
+  PKCS7_RECIP_INFO *ri = NULL;
+  ASN1_OCTET_STRING *content = NULL;
+
+  // The content field in the PKCS7 ContentInfo is optional, but that only
+  // applies to inner content (precisely, detached signatures). When reading
+  // content, missing outer content is therefore treated as an error. When
+  // creating content, PKCS7_content_new() must be called before calling this
+  // method, so a NULL p7->d is always an error.
+  if (p7->d.ptr == NULL) {
+    OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_NO_CONTENT);
+    return NULL;
+  }
+
+  switch (OBJ_obj2nid(p7->type)) {
+    case NID_pkcs7_signed:
+      md_sk = p7->d.sign->md_algs;
+      content = PKCS7_get_octet_string(p7->d.sign->contents);
+      break;
+    case NID_pkcs7_signedAndEnveloped:
+      md_sk = p7->d.signed_and_enveloped->md_algs;
+      rsk = p7->d.signed_and_enveloped->recipientinfo;
+      xalg = p7->d.signed_and_enveloped->enc_data->algorithm;
+      evp_cipher = p7->d.signed_and_enveloped->enc_data->cipher;
+      if (evp_cipher == NULL) {
+        OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_CIPHER_NOT_INITIALIZED);
+        goto err;
+      }
+      break;
+    case NID_pkcs7_enveloped:
+      rsk = p7->d.enveloped->recipientinfo;
+      xalg = p7->d.enveloped->enc_data->algorithm;
+      evp_cipher = p7->d.enveloped->enc_data->cipher;
+      if (evp_cipher == NULL) {
+        OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_CIPHER_NOT_INITIALIZED);
+        goto err;
+      }
+      break;
+    case NID_pkcs7_digest:
+      content = PKCS7_get_octet_string(p7->d.digest->contents);
+      if (!pkcs7_bio_add_digest(&out, p7->d.digest->digest_alg)) {
+        goto err;
+      }
+      break;
+    case NID_pkcs7_data:
+      break;
+    default:
+      OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_UNSUPPORTED_CONTENT_TYPE);
+      goto err;
+  }
+
+
+  for (size_t i = 0; i < sk_X509_ALGOR_num(md_sk); i++) {
+    if (!pkcs7_bio_add_digest(&out, sk_X509_ALGOR_value(md_sk, i))) {
+      goto err;
+    }
+  }
+
+  if (evp_cipher != NULL) {
+    unsigned char key[EVP_MAX_KEY_LENGTH];
+    unsigned char iv[EVP_MAX_IV_LENGTH];
+    int keylen, ivlen;
+    EVP_CIPHER_CTX *ctx;
+
+    if ((btmp = BIO_new(BIO_f_cipher())) == NULL) {
+      OPENSSL_PUT_ERROR(PKCS7, ERR_R_BIO_LIB);
+      goto err;
+    }
+    if (!BIO_get_cipher_ctx(btmp, &ctx)) {
+      OPENSSL_PUT_ERROR(PKCS7, ERR_R_BIO_LIB);
+      goto err;
+    }
+    keylen = EVP_CIPHER_key_length(evp_cipher);
+    ivlen = EVP_CIPHER_iv_length(evp_cipher);
+    ASN1_OBJECT_free(xalg->algorithm);
+    xalg->algorithm = OBJ_nid2obj(EVP_CIPHER_nid(evp_cipher));
+    if (ivlen > 0) {
+      RAND_bytes(iv, ivlen);
+    }
+    if (keylen > 0) {
+      RAND_bytes(key, keylen);
+    }
```

Comment:
As long as this is only an AES/ChaCha20 key this should be ok. Other symmetric key types could potentially(?) require a special generation/verification process.